### PR TITLE
[T5] enable T5 fp16

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR enables fp16 for T5 models, by clamping hidden states to the max value of the current data type.

As detailed in #9295, T5 produces large (`inf`) activations at 3 places
1. Output of `T5LayerFF`
2. Output of `T5LayerSelfAttention`
3. Output of `T5LayerCrossAttention`

To avoid these `inf` activations this PR clamps the `hidden_states` after above 3 outputs

### CODE REVIEW

**Comment by exelents** [comment]
Hello, Suraj.
Thank you for your solution. I had problems finetuning T5 network in fp16 mode. Here #9490 I described that problem was in model output I received a lot of feature tensors goes to be NaN and a loss value also became NaN.

When I installed transformers library from your branch, it seems that most of output features tensors like hidden_states are became a valid float values, but loss value still has NaN on many training steps. In my issue #9490 I published my training code and dataset files which I used for fp16 training mode, could you suggest something to repair this problem?

I looked at your changes, and tried to apply same clamping operations, as you proposed,  to loss value, but it still generates a lot of NaNs.

**Comment by patil-suraj** [comment]
hi @exelents ,
could you maybe share your training script/colab so I can take look?

**Comment by exelents** [comment]
Thank you for response. Here is my first attempt:
https://gist.github.com/exelents/ebbb9b0e560d9b916532a0adce16bbcb

**Comment by exelents** [comment]
I don't know what is my problem. I even tried solution made  by @stas00 in #9211 but I still have the same problem. 
Maybe problem is I built Pytorch from source and forgot some option? I did it because pypi's version don't support Cuda 11.2 and supported cuda 11.0 don't support my gpu (rtx 3090)?

**Comment by stas00** [comment]
I use pytorch-nightly w/ cuda-11.0 which works with rtx-3090:
```
pip install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html -U
```

pt-nightly w/ cuda-11.2 should be released really soon now. You can track it here https://github.com/pytorch/pytorch/issues/50232

**Comment by exelents** [comment]
> I use pytorch-nightly w/ cuda-11.0 which works with rtx-3090:
> 
> ```
> pip install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html -U
> ```
> 
> pt-nightly w/ cuda-11.2 should be released really soon now. You can track it here [pytorch/pytorch#50232](https://github.com/pytorch/pytorch/issues/50232)

It doesn't work. Pytorch nighty gets an error:
`nvcc fatal   : Unsupported gpu architecture 'compute_86`
That's mean that Cuda 11.0 doesn't support RTX 3090

**Comment by stas00** [comment]
I have been using it for the last 2 months, so yes it works. Chances are that you have more than one pytorch installed and you have the non-nightly version loaded, check your:
```
print(torch.__version__)
```

**Comment by exelents** [comment]
> I have been using it for the last 2 months, so yes it works. pt-1.7.1 doesn't work.
> 
> Chances are that you have more than one pytorch installed and you have the non-nightly version loaded, check your:
> 
> ```
> print(torch.__version__)
> ```
here is installed version: 1.8.0.dev20210109+cu110

> For building extensions like deepspeed, and apex and fairscale I use cuda-11.1
Okay, I'll try cuda 11.1, maybe it'll help.

**Comment by stas00** [comment]
> here is installed version: 1.8.0.dev20210109+cu110

Excellent. so what exactly do you do when you get that error?

I need to update my nightly, I'm at about dec20

**Comment by stas00** [comment]
So I just did:
```
pip install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html -U
```
Then inside deepspeed github clone:
```
rm -rf build
TORCH_CUDA_ARCH_LIST="6.1;8.6" DS_BUILD_OPS=1 pip install --no-cache -v --disable-pip-version-check -e .
```
You can remove 6.1, this is just my 2nd card.

Now inside the deepspeed PR brach, I run against a small t5 model:
```
export BS=20; rm -r output_dir; CUDA_VISIBLE_DEVICES=0,1 PYTHONPATH=../../src USE_TF=0 deepspeed --num_gpus=2 ./finetune_trainer.py --model_name_or_path t5-small --output_dir output_dir --adam_eps 1e-06 --data_dir wmt_en_ro --do_eval --do_predict --do_train --evaluation_strategy=steps --freeze_embeds --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 1000 --max_source_length 128 --max_target_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size $BS --per_device_train_batch_size $BS --predict_with_generate --eval_steps 25000  --sortish_sampler --task translation_en_to_ro --test_max_target_length 128 --val_max_target_length 128 --warmup_steps 5 --n_train 100 --n_val 100 --n_test 100 --deepspeed ds_config.json --fp16 --save_steps 1
```

All works.

**Comment by exelents** [comment]
> Excellent. so what exactly do you do when you get that error?

I change current cuda in my system to 11.0 version (cuda-toolkit-11-0 package in Ubuntu)
Then I install latest pytorch nighty by a command which you propose.
Latest, I run deepspeed training script proposed in  #9211 issue. But in parameters there I change model name to t5-large and remove language parameters from parameters fed to model

This code in utils.py:
```
        if data_args.src_lang is not None:
            self.dataset_kwargs["src_lang"] = data_args.src_lang
        if data_args.tgt_lang is not None:
            self.dataset_kwargs["tgt_lang"] = data_args.tgt_lang
```
This way I get a minimal training code that should run T5-large. It don't take my dataset like in two examples I have shown before, but it should work. What I see is only error "platform not supported" on pt nighty build, or NaNs in output tensors in version which I installed from source.

Here is training script:
https://gist.github.com/exelents/9dd3e6dec64dc0d640b85a7e0cfa53e9

**Comment by stas00** [comment]
I think the difference is that you need cuda-11.1 and not cuda-11.0 system-wide. This is where our setups diverge I think.

Careful though, there is 11.2 out there. I'm on ubuntu, I'm not at all sure it'd work w/ pt-nightly, that's why I'm not upgrading mine.

for rtx-3090 to work
- tf requires cuda-11.1
- pt works with cuda-11.0
- pt extensions need cuda-11.1 (apex, deepspeed, fairscale)

**Comment by patil-suraj** [comment]
Would be nice if we continue this discussion in the issue rather than PR :-)

**Comment by stas00** [comment]
> Would be nice if we continue this discussion in the issue rather than PR :-)

Mea culpa, you're right, @patil-suraj! I deleted my answers and moved them to https://github.com/huggingface/transformers/issues/9490 

@exelents, if possible move/delete yours too so that @patil-suraj could get a proper review for this PR.  (just edit and copy-n-paste and then delete)

Thank you!

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -640,6 +640,11 @@ def forward(
         hidden_states, present_key_value_state = self_attention_outputs[:2]
         attention_outputs = self_attention_outputs[2:]  # Keep self-attention outputs and relative position weights
 
+        # clamp inf values
+        if torch.isinf(hidden_states).any():
+            clamp_value = torch.finfo(hidden_states.dtype).max - 1000
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:645]: why the -1000?

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -640,6 +640,11 @@ def forward(
         hidden_states, present_key_value_state = self_attention_outputs[:2]
         attention_outputs = self_attention_outputs[2:]  # Keep self-attention outputs and relative position weights
 
+        # clamp inf values
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:643]: maybe improve comment slightly:

```suggestion
        # clamp inf values to enable fp16 training
```

**Comment by patrickvonplaten** [comment]
This is great! It would be awesome if you could also publish a notebook showcasing the results obtained with `google/t5v1_1-base` and `t5-base` on CNN/DailyMail with fp16

**Comment by exelents** [comment]
Dear @patil-suraj 
Your PR works well for t5 model, thank you for your work.
But now I tried new t5 model version released recently by Google: google/t5-v1_1-xl
The same code after loading google/t5-v1_1-xl instead of t5-3b is going to return a lot "overflow" errors.

Can you tell me, should your code fix fp16 on google/t5-v1_1-xl model? 
Here is training code:
https://github.com/exelents/try_t5_qa
Run ./run-qa-3b.sh

File: src/transformers/models/t5/modeling_t5.py (hunk 1 above)
Comment (patil-suraj) [reply src/transformers/models/t5/modeling_t5.py:645]: Just to be on the safer side, setting it to the exact max value might again lead to `inf` values in subsequent layers

Comment (patrickvonplaten) [reply src/transformers/models/t5/modeling_t5.py:645]: Okey just noticed that we do the same in Bart as well

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
> Dear @patil-suraj
> Your PR works well for t5 model, thank you for your work.
> But now I tried new t5 model version released recently by Google: google/t5-v1_1-xl
> The same code after loading google/t5-v1_1-xl instead of t5-3b is going to return a lot "overflow" errors.
> 
> Can you tell me, should your code fix fp16 on google/t5-v1_1-xl model?
> Here is training code:
> https://github.com/exelents/try_t5_qa
> Run ./run-qa-3b.sh
> 
> Upd: I run my code on Transformers's branch from your current PR #9487 merged with PR #9211 needed for deepspeed integration.
> Can you confirm a problem, or it's just mine?

Hey @exelents, can you include a code snippet to reproduce your error as well as the full stack trace of your error?

**Comment by patil-suraj** [comment]
@patrickvonplaten , @exelents 

as stated in #9432 

This fix works for following models and versions, with apex `01` and `native amp`
- T5v1: t5-small, t5-base, t5-large
- T5v1_1: google/t5-v1_1-small, google/t5-v1_1-base
- MT5: google/mt5-small, google/mt5-base

Just did a small experiment with `t5-v1_1-large` and it still gives `nan` loss after 200 steps, so might not work for `xl`, 

also, @exelents  by overflow error do you mean the gradient overflow warning thrown by `apex` ?

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten , @exelents
> 
> as stated in #9432
> 
> This fix works for following models and versions, with apex `01` and `native amp`
> 
> * T5v1: t5-small, t5-base, t5-large
> * T5v1_1: google/t5-v1_1-small, google/t5-v1_1-base
> * MT5: google/mt5-small, google/mt5-base
> 
> Just did a small experiment with `t5-v1_1-large` and it still gives `nan` loss after 200 steps, so might not work for `xl`,
> 
> also, @exelents by overflow error do you mean the gradient overflow warning thrown by `apex` ?

Ah ok, we still see `nan's` with `t5-v1_1-large` then :-/ Do you think this could be fixed by adding one more clamp statement? @patil-suraj

**Comment by exelents** [comment]
> Hey @exelents, can you include a code snippet to reproduce your error as well as the full stack trace of your error?
My code is here:
https://github.com/exelents/try_t5_qa
It requires deepspeed to run, as well as code from #9211 PR (deepspeed integration) be merged. Use run-qa-3b.sh to test.

Here is error stack:
https://gist.github.com/exelents/10f1d03e61059ddf2dfba7068114c93a
Look at the end - we have a message after every step:
`[2021-01-11 16:58:18,163] [INFO] [stage2.py:1361:step] [deepscale] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 256.0, reducing to 128.0`

**Comment by patil-suraj** [comment]
> Do you think this could be fixed by adding one more clamp statement?

I'm again trying to locate where exactly in the model this happens, in case it's the same as above (first `inf` then `nan` ) then we could fix it by adding one more clamp

**Comment by exelents** [comment]
I have checked a loss value, and it seems in is not NaN. It got values like "48.7500" or "40.9688" but there are vaild values. Despite that I see messages like "OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 1024.0, reducing to 512.0", that it seems means that something bad happened with model's loss.

**Comment by sgugger** [comment]
> Attempted loss scale: 1024.0, reducing to 512.0", that it seems means that something bad happened with model's loss.

Those warnings don't mean anything went wrong, it's logical with dynamic loss scaling that some loss scale values are too big at the beginning of training.

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing this!

**Review by LysandreJik — APPROVED** [review]
Very cool! Thanks for working on this @patil-suraj!
